graphbolt/src/csc_sampling_graph.cc

```diff
@@ -207,6 +198,34 @@ c10::intrusive_ptr<SampledSubgraph> CSCSamplingGraph::SampleNeighbors(
       subgraph_reverse_edge_ids, subgraph_type_per_edge);
 }
 
+c10::intrusive_ptr<SampledSubgraph> CSCSamplingGraph::SampleNeighbors(
+    const torch::Tensor& nodes, const std::vector<int64_t>& fanouts,
+    bool replace, bool labor, bool return_eids,
+    torch::optional<std::string> probs_name) const {
+  torch::optional<torch::Tensor> probs_or_mask = torch::nullopt;
+  if (probs_name.has_value() && !probs_name.value().empty()) {
+    probs_or_mask = edge_attributes_.value().at(probs_name.value());
+    // Note probs will be passed as input for 'torch.multinomial' in deeper
+    // stack, which doesn't support 'torch.half' and 'torch.bool' data types. To
+    // avoid crashes, convert 'probs_or_mask' to 'float32' data type.
+    if (probs_or_mask.value().dtype() == torch::kBool ||
+        probs_or_mask.value().dtype() == torch::kFloat16) {
+      probs_or_mask = probs_or_mask.value().to(torch::kFloat32);
+    }
+  }
+  if (labor) {
+    TORCH_CHECK(
+        !replace, "Sampling with replacement is not supported for labor.");
```

Comment:
There are 2 concerns here:
1. labor sampling is not yet a widely received sampling algorithm, it might be confusing to user who have no clue about what's labor.
2. The code here are intertwined heavily, creating a lot of if-else, which is not very readability friendly.

I'd suggest:
PR#1. Create a `def sample_neighbors_labor(...)`, put a clear documentation without implementation and add empty unit tests.

PR#2. Create a new C++ implementation, instead of overwriting SampleNeighbors, create SampleNeighborsLabor(...), same for pick function, and update the unit tests coorespondingly.

Reply:
So we don't mind copying the whole of Sample Neighbors code along with PickByEType binary search logic? If I separate out the labor code, it will be an extra maintenance burden and fall behind the neighbor sampler when it comes to functionalities as the neighbor sampler code is updated. This code strictly aims to be %100 compatible with NeighborSampling. Even the existing tests can't detect if labor is enabled, they all just pass.

My plan was to have another version with a separate API where the %100 compatibility is not followed (Implementing the expected fanout algorithm with importance sampling as in dgl.sampling.sample_labors).

About 1: It is possible to change the bool parameter's name to layer, since layer sampling algorithms have been around for 5+ years, users should be more familiar with them. We can explain in the documentation that what enables layer sampling functionality for NeighborSampler is the Labor algorithm.

About 2: Would it be acceptable if I made the code more readable by eliminating the extra ifs but still using templates to share the code? It is possible to do that. I just want this code to be maintained just as well as neighbor sampling.

Reply:
I will shortly push a much more readable version to illustrate what I mean.

Reply:
I had to use only 1 extra if expression potentially making it less readable, and it is to dispatch Neighbor or Labor from SampleNeighbors by passing the right template parameter to SampleNeighborsImpl. The code should read the same now, the other existing functions are untouched except that they are templated now. What do you think?

I also changed the outer looking API parameter name to layer which the users should already know about.

Reply:
My reasoning is that if somebody for example decided to optimize the multithreading logic for neighbor sampling, the benefits will automatically transfer to labor as well, the goal is to keep their quality on par so that labor is always production ready too.

Reply:
Thanks for the quick action, the implementation now is much more readable. I made some comment inline.

For the python level API, let me discuss with my team to see whether people prefer keep a single api with option or keep separate apis.

Reply:
The name of the labor paper is Layer-Neighbor Sampling -- Defusing Neighborhood Explosion in GNNs. So it is Neighbor Sampling with layer sampling :). So it would factually be correct for SampleNeighbors to have a layer bool parameter. But, of course, it is your and your team's decision.